skyportal/handlers/api/earthquake.py

```diff
@@ -0,0 +1,616 @@
+from obspy.taup.helper_classes import TauModelError
+import sqlalchemy as sa
+from sqlalchemy.orm import joinedload
+from sqlalchemy.orm import sessionmaker, scoped_session
+import obspy
+
+from baselayer.app.access import auth_or_token
+from baselayer.log import make_log
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    EarthquakeEvent,
+    EarthquakeNotice,
+    EarthquakePrediction,
+    MMADetector,
+    User,
+)
+
+log = make_log('earthquake')
+
+
+Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+
+
+def post_earthquake_from_xml(payload, user_id, session):
+    """Post Earthquake to database from quakeml xml.
+    payload: str
+         readable string
+    user_id : int
+        SkyPortal ID of User posting the Earthquake
+    session: sqlalchemy.Session
+        Database session for this transaction
+    """
+
+    user = session.query(User).get(user_id)
+
+    event = obspy.core.event.read_events(io.StringIO(payload), format="QUAKEML")[0]
+    event_uri = event.resource_id.id.replace("/", "-")
+
+    split_strings = ["1-query?eventid=", "-EVENT-", "-Event-", "-event-", "-geofon-"]
+    event_id = None
+    for split_string in split_strings:
+        str_split = event_uri.split(split_string)
+        if len(str_split) > 1:
+            event_id = str_split[-1]
+            break
+    if event_id is None:
+        event_id = event_uri
+
+    event_type = event.event_type
+    if event_type == "not existing":
+        event = session.scalars(
+            EarthquakeEvent.select(user, mode="update").where(
+                EarthquakeEvent.event_id == event_id
+            )
+        ).first()
+        if event is not None:
+            event.status = 'canceled'
+            session.add(event)
+            session.commit()
+
+            return event.id
+        else:
+            return None
+
+    magnitudes = event.magnitudes
+    if len(magnitudes) == 0:
+        raise ValueError('Must have magnitude information to create Earthquake.')
+
+    mag = magnitudes[-1].mag
+    origin = event.origins[-1]
+
+    event = session.scalars(
+        EarthquakeEvent.select(user).where(EarthquakeEvent.event_id == event_id)
+    ).first()
+
+    if event is None:
+        event = EarthquakeEvent(
+            event_id=event_id, event_uri=event_uri, sent_by_id=user.id
+        )
+        session.add(event)
+    else:
+        if not event.is_accessible_by(user, mode="update"):
+            raise ValueError(
+                "Insufficient permissions: Earthquake event can only be updated by original poster"
+            )
+
+    earthquake_notice = EarthquakeNotice(
+        content=payload.encode('utf-8'),
+        event_id=event_id,
+        lat=origin.latitude,
+        lon=origin.longitude,
+        depth=origin.depth,
+        magnitude=mag,
+        date=origin.time.datetime,
+        sent_by_id=user.id,
+    )
+    session.add(earthquake_notice)
+    session.commit()
```

Comment:
I think that unless you get a "verified" session from `self.Session` of the handler, this `commit` does not check if the user has permissions to add things.

Reply:
I guess you'd have to explicitly call `is_accessible` with the 'create' option and the user.

Reply:
ok that's a nifty trick.